src/nunit.analyzers/EqualToIncompatibleTypes/EqualToIncompatibleTypesAnalyzer.cs

```diff
@@ -0,0 +1,268 @@
+                if (namedActualType.IsTupleType && namedExpectedType.IsTupleType)
+                {
+                    return namedActualType.TupleElements.Length == namedActualType.TupleElements.Length
+                        && Enumerable.Range(0, namedActualType.TupleElements.Length).All(i =>
+                            CanBeAssertedForEquality(namedActualType.TupleElements[i].Type, namedExpectedType.TupleElements[i].Type,
+                                semanticModel, checkedTypes));
+                }
+
+                // Dictionaries
+                if (IsDictionary(namedActualType, actualFullName, out var actualKeyType, out var actualValueType)
+                    && IsDictionary(namedExpectedType, expectedFullName, out var expectedKeyType, out var expectedValueType))
+                {
+                    // Unlike for KeyValuePairs, Dictionaries Keys should match exactly
+
+                    var keysConversion = semanticModel.Compilation.ClassifyConversion(actualKeyType, expectedKeyType);
+                    var keysMatching = keysConversion.IsIdentity || keysConversion.IsReference;
+
+                    return keysMatching && CanBeAssertedForEquality(actualValueType, expectedValueType, semanticModel, checkedTypes);
+                }
+            }
+
+            // IEnumerables
+            if (actualType.IsIEnumerable(out var actualElementType) && expectedType.IsIEnumerable(out var expectedElementType))
+            {
+                if (actualElementType == null || expectedElementType == null)
+                {
+                    // If actual or expected values implement only non-generic IEnumerable, we cannot determine
+                    // whether types are suitable
+                    return true;
+                }
+                else
+                {
+                    return CanBeAssertedForEquality(actualElementType, expectedElementType, semanticModel, checkedTypes);
+                }
+            }
+
+            if (namedActualType != null && namedExpectedType != null)
+            {
+                // KeyValuePairs
+                if (IsKeyValuePair(namedActualType, actualFullName, out var actualKeyType, out var actualValueType)
+                    && IsKeyValuePair(namedExpectedType, expectedFullName, out var expectedKeyType, out var expectedValueType))
+                {
+                    return CanBeAssertedForEquality(actualKeyType, expectedKeyType, semanticModel, checkedTypes)
+                        && CanBeAssertedForEquality(actualValueType, expectedValueType, semanticModel, checkedTypes);
+                }
+
+                // Streams
+                if (IsStream(actualType, actualFullName) && IsStream(expectedType, expectedFullName))
+                    return true;
+
+                // IEquatables
+                if (IsIEquatable(actualType, expectedType) || IsIEquatable(expectedType, actualType))
+                    return true;
+            }
```

Comment:
Should we move the content of this if-statement up inside the `if (namedActualType != null && namedExpectedType != null)` expression on L127, so we don't have this minimal duplication?

Reply:
I tried to preserve same order as it goes in constraint itself, but probably KeyValuePair case is order independent

Reply:
Ok. Then I think it is to leave it as it is.